### PR TITLE
issue #779 #1150 partitionize bulkexport JavaBatch Job & fix CWWKY0041W warnings

### PR DESCRIPTION
**Pull request by albertwang-ibm**
**Summary of the changes:**
(1) Partitionized bulkdata export jobs - system, patient and group exports.
(2) Dropped single COS file multi-parts uploading for the whole export job.
(3) Added multi-parts uploading for each partition.
(4) Added server configurations to control the file size or resource number of each COS file.
      - fhirServer/bulkdata/cosFileMaxResources
      - fhirServer/bulkdata/cosFileMaxSize
(5) Added simple metrics for the exporting jobs.
(6) Fixed CWWKY0041W warnings of bulkdata both export and import jobs.
(7) Refactoring of the bulkdata export codes.

### CODE REVIEW

**Comment by prb112** [comment]
```
   symbol: class Dependent
[ERROR] /home/runner/work/FHIR/FHIR/fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ExportPartitionCollector.java:[14,32] package javax.enterprise.context does not exist
[ERROR] /home/runner/work/FHIR/FHIR/fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ExportPartitionCollector.java:[20,2] cannot find symbol
```

Needs a dependency like 

```
 <dependency>
            <groupId>javax</groupId>
            <artifactId>javaee-api</artifactId>
            <scope>provided</scope>
        </dependency>
```

note there is no version as it should come from the parent.

File: fhir-parent/pom.xml [hunk 1]
```diff
@@ -297,6 +297,12 @@
                 <artifactId>jakarta.inject-api</artifactId>
                 <version>1.0</version>
             </dependency>
+            <dependency>
+                <groupId>javax.enterprise</groupId>
+                <artifactId>cdi-api</artifactId>
+                <version>2.0.SP1</version>
+                <scope>provided</scope>
```
Comment (prb112) [inline fhir-parent/pom.xml:304]: Please use

<dependency>
    <groupId>jakarta.enterprise</groupId>
    <artifactId>jakarta.enterprise.cdi-api</artifactId>
    <version>2.0.2</version>
    <scope>provided</scope>
</dependency>

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 2]
```diff
@@ -1486,6 +1486,8 @@ This section contains reference information about each of the configuration prop
 |`fhirServer/bulkdata/validBaseUrls`|string|The list of supported urls which are approved for the fhir server to access|
 |`fhirServer/bulkdata/validBaseUrlsDisabled`|boolean|Disables the URL checking feature|
 |`fhirServer/bulkdata/maxInputPerRequest`|integer|The maximum inputs per bulk import|
+|`fhirServer/bulkdata/cosFileMaxResources`|int|Max FHIR resources per COS file, "-1" means no limit |
```
Comment (prb112) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1489]:
```suggestion
|`fhirServer/bulkdata/cosFileMaxResources`|int|The maximum number of FHIR resources per COS file, "-1" means no limit |
```

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 3]
```diff
@@ -1486,6 +1486,8 @@ This section contains reference information about each of the configuration prop
...
+|`fhirServer/bulkdata/cosFileMaxSize`|int|Max COS file size in bytes, "-1" means no limit |
```
Comment (prb112) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1490]:
```suggestion
|`fhirServer/bulkdata/cosFileMaxSize`|int|The maximum COS file size in bytes, "-1" means no limit |
```

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 4]
```diff
@@ -1532,6 +1534,8 @@ This section contains reference information about each of the configuration prop
 |`fhirServer/audit/serviceProperties/geoCounty`|US|
 |`fhirServer/bulkdata/isExportPublic`|true|
 |`fhirServer/bulkdata/validBaseUrlsDisabled`|false|
+|`fhirServer/bulkdata/cosFileMaxResources`|500000|
+|`fhirServer/bulkdata/cosFileMaxSize`|209715200|
```
Comment (prb112) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1538]: What does this equate to? 2G?

File: docs/src/pages/guides/FHIRServerUsersGuide.md [hunk 5]
```diff
@@ -1596,6 +1600,8 @@ must restart the server for that change to take effect.
 |`fhirServer/bulkdata/validBaseUrls`|Y|Y|
 |`fhirServer/bulkdata/maxInputPerRequest`|Y|Y|
 |`fhirServer/bulkdata/validBaseUrlsDisabled`|Y|Y|
+|`fhirServer/bulkdata/cosFileMaxResources`|N|Y|
+|`fhirServer/bulkdata/cosFileMaxSize`|N|Y|
```
Comment (prb112) [inline docs/src/pages/guides/FHIRServerUsersGuide.md:1604]: should this be server wide?

File: fhir-bulkimportexport-webapp/pom.xml [hunk 6]
```diff
@@ -33,6 +33,10 @@
             <groupId>org.glassfish</groupId>
             <artifactId>jakarta.json</artifactId>
         </dependency>
+        <dependency>
+            <groupId>javax.enterprise</groupId>
+            <artifactId>cdi-api</artifactId>
+        </dependency>
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/pom.xml:39]:
```
<dependency>
    <groupId>jakarta.enterprise</groupId>
    <artifactId>jakarta.enterprise.cdi-api</artifactId>
    <version>2.0.2</version>
    <scope>provided</scope>
</dependency>
```

File: fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml [hunk 7]
```diff
@@ -1,20 +1,23 @@
...
+    </listeners>
     <step id="step1">
-        <chunk checkpoint-policy="item" item-count="1">
+        <chunk checkpoint-policy="custom">
             <reader ref="com.ibm.fhir.bulkexport.group.ChunkReader">
                 <properties >
                     <property name="fhir.tenant" value="#{jobParameters['fhir.tenant']}"/>
                     <property name="fhir.datastoreid" value="#{jobParameters['fhir.datastoreid']}"/>                   
-                    <property name="fhir.resourcetype" value="#{jobParameters['fhir.resourcetype']}"/>
+                    <property name="partition.resourcetype" value="#{partitionPlan['partition.resourcetype']}"/>
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml:12]: Why the change from the fhir. prefix?

File: fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml [hunk 8]
```diff
@@ -23,9 +26,24 @@
                     <property name="cos.credential.ibm" value="#{jobParameters['cos.credential.ibm']}"/>
                     <property name="cos.bucket.name" value="#{jobParameters['cos.bucket.name']}"/>
                     <property name="cos.bucket.pathprefix" value="#{jobParameters['cos.bucket.pathprefix']}"/>
-                    <property name="fhir.resourcetype" value="#{jobParameters['fhir.resourcetype']}"/>
+                    <property name="partition.resourcetype" value="#{partitionPlan['partition.resourcetype']}"/>
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml:29]: Why the change from the fhir. prefix?

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/common/CheckPointAlgorithm.java [hunk 9]
```diff
@@ -6,38 +6,45 @@
...
+    @BatchProperty(name = Constants.COS_BUCKET_FILE_MAX_RESOURCES)
+    String cosBucketFileMaxResources;
 
     /**
      * The file size limit when exporting to multiple COS files.
      */
     @Inject
-    @BatchProperty(name = "cos.bucket.maxfilesize")
-    String cosBucketMaxFileSize;
+    @BatchProperty(name = Constants.COS_BUCKET_FILE_MAX_SZIE)
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/common/CheckPointAlgorithm.java:46]:
```suggestion
    @BatchProperty(name = Constants.COS_BUCKET_FILE_MAX_SIZE)
```

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/Constants.java [hunk 10]
```diff
@@ -40,17 +40,26 @@
     public static final String COS_LOCATION = "cos.location";
     public static final String COS_BUCKET_NAME = "cos.bucket.name";
     public static final String COS_IS_IBM_CREDENTIAL = "cos.credential.ibm";
+    public static final String COS_BUCKET_FILE_MAX_SZIE = "cos.bucket.filemaxsize";
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/Constants.java:43]:
```suggestion
    public static final String COS_BUCKET_FILE_MAX_SIZE = "cos.bucket.filemaxsize";
```

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java [hunk 11]
```diff
@@ -41,23 +42,22 @@
...
 public class ChunkReader extends AbstractItemReader {
     private final static Logger logger = Logger.getLogger(ChunkReader.class.getName());
     protected int pageNum = 1;
-    protected int indexOfCurrentResourceType = 0;
     // Control the number of records to read in each "item".
     protected int pageSize = Constants.DEFAULT_SEARCH_PAGE_SIZE;
 
     protected FHIRPersistence fhirPersistence;
-    protected List<String> resourceTypes;
+    Class<? extends Resource> resourceType;
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java:60]: protected? 
this change from List is because each resourceType is split into separate executions?

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ExportJobListener.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        int totalExportedFhirResources = 0;
+        List<String> resourceTypeSummaries = new ArrayList<>();
+        for (CheckPointUserData partitionSummary : partitionSummaries) {
+            logger.info(partitionSummary.getResourceTypeSummary() + "\t|"
+                  + partitionSummary.getTotalResourcesNum());
+            resourceTypeSummaries.add(partitionSummary.getResourceTypeSummary());
+            totalExportedFhirResources += partitionSummary.getTotalResourcesNum();
+        }
+
+        logger.info(" ---- Total: " + totalExportedFhirResources
```
Comment (prb112) [inline fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ExportJobListener.java:86]: this seems like a good thing to have a log.fine

File: fhir-parent/pom.xml (hunk 1 above)
Comment (albertwang-ibm) [reply fhir-parent/pom.xml:304]: good point, will change

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 4 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1538]: 200M

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 5 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1604]: this is server wide at present.

File: fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml (hunk 7 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml:12]: this is for the resourcetype processed by the partition(passed by the partition mapper to the chunk codes), this is different from the job parameter.

File: fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml (hunk 8 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/META-INF/batch-jobs/FhirBulkExportGroupChunkJob.xml:29]: answer as above.

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/common/CheckPointAlgorithm.java (hunk 9 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/common/CheckPointAlgorithm.java:46]: ugly typo error, let me fix. :)

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/Constants.java (hunk 10 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/Constants.java:43]: good catch

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java (hunk 11 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/patient/ChunkReader.java:60]: yes, not each partition/chuck codes only need to process its own resouretype - one and only one resource type.

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 5 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1604]: Right, is this something that should be configured by the tenant

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 4 above)
Comment (prb112) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1538]: I think it should be added to the documentation (not necessarily in this row), that it equates to 200M

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ExportJobListener.java (hunk 12 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/system/ExportJobListener.java:86]: this is the summary of the whole export job, so I used info instead of fine.

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 4 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1538]: let me add to the comment for this config. after -1.

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 5 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1604]: this is the same as isExportPublic config, they are all server wide at present. we have another issue to make them per tenant.

File: fhir-parent/pom.xml (hunk 1 above)
Comment (albertwang-ibm) [reply fhir-parent/pom.xml:304]: done

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 3 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md]: done

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 4 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1538]: done

File: fhir-bulkimportexport-webapp/pom.xml (hunk 6 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/pom.xml:39]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/common/CheckPointAlgorithm.java (hunk 9 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkexport/common/CheckPointAlgorithm.java]: done

File: fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/Constants.java (hunk 10 above)
Comment (albertwang-ibm) [reply fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/bulkcommon/Constants.java]: done

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 2 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md]: done

File: docs/src/pages/guides/FHIRServerUsersGuide.md (hunk 5 above)
Comment (albertwang-ibm) [reply docs/src/pages/guides/FHIRServerUsersGuide.md:1604]: issue #943
